components/raftstore/src/store/fsm/store.rs

```diff
@@ -1544,6 +1511,47 @@ impl<'a, T: Transport, C: PdClient> StoreFsmDelegate<'a, T, C> {
         region_id: u64,
         msg: &RaftMessage,
         is_local_first: bool,
+    ) -> Result<bool> {
+        if !is_initial_msg(msg.get_message()) {
+            let msg_type = msg.get_message().get_msg_type();
+            debug!(
+                "target peer doesn't exist, stale message";
+                "target_peer" => ?msg.get_to_peer(),
+                "region_id" => region_id,
+                "msg_type" => ?msg_type,
+            );
+            self.ctx.raft_metrics.message_dropped.stale_msg += 1;
+            return Ok(false);
+        }
+
+        if is_local_first {
+            let mut pending_create_peers = self.ctx.pending_create_peers.lock().unwrap();
+            if pending_create_peers.contains_key(&region_id) {
+                return Ok(false);
+            }
+            pending_create_peers.insert(region_id, (msg.get_to_peer().get_id(), false));
+        }
+
+        let res = self.maybe_create_peer_internal(region_id, &msg, is_local_first);
+        // If failed, i.e. Err or Ok(false), remove this peer data from `pending_create_peers`.
+        if !*res.as_ref().unwrap_or(&false) {
```

Comment:
How about `if res.as_ref().map_or(true, |b| !*b) && is_local_first {`? `*` is preferred than `&`, less level of negatives is preferred.